### DISCUSSION TITLE
CUDF support

### DISCUSSION THREAD

**Discussion by hugosenari** [nix-community/dream2nix#418] [Ideas]
Hi,

I'm currently creating an alternative for [flake-nimble](/nix-community/flake-nimble).

My first attempt was let nix flakes locks handle locks, it didn't work well.
My second attempt was, use lock tool from ecosystem, but it is too slow.
And the current approach is indexing all and use head, isn't ideal, so I'm looking for alternatives to dependency resolution.

I was looking for SAT/SMT solutions, and found [aspcud](/potassco/aspcud/) it uses [cudf](https://www.mancoosi.org/cudf/) to resolve dependencies.
And I'm wondering if it makes sense to add CUDF support to dream2nix.

I'm also open to ear about other options, like does dream2nix solve this?

**Comment by DavHau** [comment]
So far we got around dealing with dependency resolution in general.
We either read from already existing lock files (pure translation) or we call out to external dependency management tools (impure resolution), to create a lock file that we can parse. The tools invoked by our impure translator do dependency resolution to create the lock file, but we treat that as a block box. For example for nodejs projects we just call out to `npm` to create a lock file if none exists.

If Nim does not have a tool yet that can create a lockfile, then implementing your own resolver might be worthwhile.
In general it would be amazing if someone added support for dependency resolution to dream2nix.
The nix ecosystem is missing such functionality in general. I wish we can have `builtins.resolve` or something similar one day.

Currently dream2nix can only act as an addon to dependency management. It relies on other tools maintaining lock files. It cannot work standalone, because there is no dependency resolver. So if you can solve that, that would be great ;)

**Reply by hugosenari** [reply]
My [POC](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/example/flake.nix#L14) now works.
Documenting here, if anyone else would like to try for different lang or the same.

Example requesting 3 packages,
hcparse has [11 dependencies](http://www.plantuml.com/plantuml/uml/xPLH3i8W38RVFGKNy0YQ7S4hC22ZEWgX9OintbrmSs5UO9ZunDylVVXI6kY8PUJaneCG6eOeWnMNso0UkLOUbK6Eais9Op8PMHLa92EEvq_yWWv85LJE0iYQSQp1bJpUQz0RL7P34eBoskY-Xawgw1F3I3MTWEtM0C5bO-3IjkOTrROEkPB4fMrEpOY5407VOdFQyp8hWuRcLMlfA2V0eT0bjh7yJSshqc9QtuebDYBsN_iVNVldeJEeyx_r0W==)
libp2p has [16 dependencies](http://www.plantuml.com/plantuml/uml/xLR1RiCW3BttA_Y17HPfkstg9-mF9aAyXeK0mewgTjg_3ogM4f8sgRPJr1EFvsTRnfQJBH9trBLgmrWbTuxRchqrE_Rbxy9e0PeS9tWXru6db2njOTd3QqG3gWLoKg3d8Wgag9tH9f0H1R8ciXr95UWU1qEd9H4W5Ra12VQU2qUkI-9Ue9Svby0TOdY8Y7BjkT9fzagQiqZBLhY39JDWRyYzoWVTGub6D9wBwD-BIk8kKwMDIkZPeXBjYbgLLoM5WhwmqoMOWlgtjXvFfkrTFj7eTmGdkP97Jj9eBncJCwLSSCZ2ZWHfY5lIZRGV70d90Mz3wSajzynCLt9aodNEqj6l0C-4O_46hJGA9Vs7QRXyGTZYwRbvFBKdujas9hhfnNmVzlkm_s7OlpTRq9N_lVe1)
cligen has 0 dependencies
```nix
dependencies = nimrevs.lib.srcs {
  inherit pkgs;
  nimPkgs = [
    "cligen"
    "hcparse"
    "libp2p"  
  ];
};
```
Can be built it as:
```bash
nix show-derivation 'github:riinr/flake-crown?dir=pkgsRev/example'
nix build 'github:riinr/flake-crown?dir=pkgsRev/example'
nix build 'github:riinr/flake-crown?dir=pkgsRev/example'#nixpkgs2211 # fail :(
```

Pkg Resolution process:
- indexing time
    1. [index metadata as cudf](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/packages.cudf)
    2. index metadata as json ([example](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/codeberg.org/annaaurora/snekim/meta.json))
    3. [index cudf <-> json](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/packages.nix), because cudf has name (no `_`) and [version (must be int)](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/codeberg.org/annaaurora/snekim/meta.json#L216) restrictions
- buildtime
    1. [call aspcud](https://github.com/riinr/flake-crown/blob/flake-pinning/pkgsRev/resolve.nix#L17) with a "request" (ie: cudf index + 'install: cligen, hcparse, libp2p >= 1')
    2. parse the response (i replaced `:` with `=` called it as TOML)
    3. use the cudf <-> json to find ref, rev and meta info
    4. fetchGit {ref, rev}

Index process (Nim specific):
1. fetch [packages.json](https://github.com/nim-lang/packages/blob/master/packages.json) (is like nixpkgs of Nim)
2. git ls-remote for every pkg
3. clone every ref  (Nim has no std file name for package meta and it is a Nim file)
4. use nimble dump (only available in latest version)
5. write the output of dump to a json file

Notes and questions:
- Not sure how it would be with large package set (Nim has only ~2k)
- Nim can't have two versions of same package, Python and Node can, how CUDF deal with it?
- There is a bug for [aspcud with large version numbers](https://github.com/potassco/aspcud/issues/11) (converting semver to int isn't lossless)
- This POC [didn't work for 22.11](https://github.com/NixOS/nixpkgs/issues/206333), I don't know why.
- I'm betting on aspcud precision, because it uses pure logic to resolve dependencies, [Opam](https://opam.ocaml.org/doc/External_solvers.html) used to recommend it. Other options? ;-)
- This POC doesn't propagate dependencies. It only downloads and adds to compiler path.

**Comment by DavHau** [comment]
Hey, I finally took a closer look at this and I'm quite hyped about this experiment.

> Not sure how it would be with large package set (Nim has only ~2k)

I'd love to find out. [pypi-deps-db](https://github.com/DavHau/pypi-deps-db) could be used as a data source. It contains most of the 4 million package releases from pypi.org.

> Nim can't have two versions of same package, Python and Node can, how CUDF deal with it?

Node can, but python cannot. The [CUDF Primer](https://www.mancoosi.org/cudf/primer/) mentions:
`conflicts are not implicit among different versions of the same package: multiple versions of the same package can be installed at the same time.
Self-conflicts are ignored. This means that you can achieve Debian-like implicit conflicts (i.e. only one version of each package can be installed at any given time) as follows: [...]`
It seems like by default CUDF allows multiple versions if the package doesn't list its own name under `conflicts`.

> There is a bug for https://github.com/potassco/aspcud/issues/11 (converting semver to int isn't lossless)

Hm, it seems like the tooling around CUDF is not super well maintained. Did you look into any alternatives so far, like for example [libsolv](https://github.com/openSUSE/libsolv) ?

**Comment by hugosenari** [comment]
> Hey, I finally took a closer look at this and I'm quite hyped about this experiment.

At least spammer called your attention :rofl:

> I'd love to find out. [pypi-deps-db](https://github.com/DavHau/pypi-deps-db) could be used as a data source. It contains most of the 4 million package releases from pypi.org.

SAT Solver geeks are proud of how fast theirs software 'solve large problem sets', I am not sure if is the case here, but there is a hope.
The nice thing, is that once solved, the resolution goes to storage and will be cached until index or requirements change, that also means we could share this cache. :star_struck: 

> Hm, it seems like the tooling around CUDF is not super well maintained

aspcud at least is in use by OCaml community, they are migration to own solution, that is why we have it in nixpkgs.
The problem with large int, I solved with two 0 padding algorithms, one for SemVer other for CalVer. From SemVer perspective, we could also remove Patch information from package and only index latest patch, or we could do like shared libraries ie: openssl 3.2.1 is openssl3 2.1, because openssl1 and openssl2 are incompatible.
The problem with nixpkgs 22.11, was also fixed.


> Did you look into any alternatives so far, like for example [libsolv](https://github.com/openSUSE/libsolv) ?

No because, CUDF + aspcud solution was too simple like
```nix
readFile (
  runCommand "resolv-deps" {
    cudfRequest = "request: \ninstall: myDependency\n";
    passAsFile  = ["cudfRequest"];
    buildInputs = [ aspcud ];
  }
  ''
    aspcud <(cat ${nimPkgsCudfIndex} $cudfRequestPath) $out
  ''
)
```

**Comment by hugosenari** [comment]
> I'd love to find out. [pypi-deps-db](https://github.com/DavHau/pypi-deps-db) could be used as a data source.

It can be used as data source.
https://github.com/DavHau/pypi-deps-db/pull/17/files
![shot-2023-01-04_07-31-53](https://user-images.githubusercontent.com/863299/210536196-289f2ed8-4560-4248-ae45-f216f0eb6f1a.jpg)

Now you need the reverse mapping.
